### ISSUE TITLE
Best approach to redirect manually and appending the UserAccessToken

### ISSUE THREAD

**Issue by greg-signi** [DuendeSoftware/Support#1050]
**Which version of Duende BFF are you using?**
2.1.1

**Which version of Yarp.ReverseProxy are you using?**
2.1.0

**Which version of .NET are you using?**
6.0

This is a question to gauge personal opinions and to make sure I didn't miss anything your framework had to offer..

This is the use case I have

_**Front-end makes Api request -> ReverseProxy (Notices Bff cookie and forwards to Bff) -> Bff -> Redirects to Api request with access token**_

So now I couldn't use your out of the box extension methods to append the access token to the redirect I was hoping to achieve, so I took inspiration in your `Duende.Bff.Yarp.RemoteApiEndpoint.Map(string localPath, string apiAddress)` and came up with this code

```
// Arrange
UserTokenRequestParameters userTokenRequestParameters;
BffUserAccessTokenParameters bffUserAccessTokenParameters = new();
userTokenRequestParameters = bffUserAccessTokenParameters.ToUserAccessTokenRequestParameters();
var result = await _accessTokenRetriever.GetAccessToken(new AccessTokenRetrievalContext
{
    HttpContext = HttpContext,
    Metadata = new BffRemoteApiEndpointMetadata
    {
        RequiredTokenType = TokenType.User,
        BffUserAccessTokenParameters = bffUserAccessTokenParameters,
    },
    UserTokenRequestParameters = userTokenRequestParameters,
});

var forwarder = HttpContext.RequestServices.GetRequiredService<IHttpForwarder>();
var clientFactory = HttpContext.RequestServices.GetRequiredService<IHttpMessageInvokerFactory>();
var transformerFactory = HttpContext.RequestServices.GetRequiredService<IHttpTransformerFactory>();

string localPath = HttpContext.Request.Path;
var httpClient = clientFactory.CreateClient(localPath);
var transformer = transformerFactory.CreateTransformer(localPath, result);

await forwarder.SendAsync(
  HttpContext,
  $"{HttpContext.Request.Scheme}://{HttpContext.Request.Host}/{destinationApiUrl}",
  httpClient,
  ForwarderRequestConfig.Empty,
  transformer);

var errorFeature = HttpContext.Features.Get<IForwarderErrorFeature>();
if (errorFeature != null)
{
    var error = errorFeature.Error;
    var exception = errorFeature.Exception;

    _logger.LogError(localPath, exception?.ToString() ?? error.ToString());
}

return new EmptyResult();
```

For this code I also have to register these services

```
services
  .AddBff()
  .AddRemoteApis(); // This
 
// Both of these  
services
   .AddReverseProxy()
   .AddBffExtensions();
```

I'm asking this because I want to make sure there's not a more **obvious & correct solution to what I'm trying to acomplish**, and to also make sure **I'm arranging the parameters correctly in the first part of the snippet..**

**Comment by josephdecock** [comment]
Can you clarify what your goal is please?

In this:

> Front-end makes Api request -> ReverseProxy (Notices Bff cookie and forwards to Bff) -> Bff -> Redirects to Api request with access token

is the final hop really a _redirect_? The yarp code looks like it is forwarding the request rather than retiring an HTTP 302 or some such.

Is the reverse proxy relevant here?

**Comment by greg-signi** [comment]
Yes my mistake, it's not a redirect but a forward instead. I would think I want a 200 OK response rather than a 302 Redirect.

The reverse proxy is mandatory since it's part of the infrastructure and every request must go by it.

**Comment by josephdecock** [comment]
Your use case then sounds pretty typical. If I'm understanding correctly, you want your BFF to proxy requests to external APIs, which is the main use case that it is designed for. The presence of a reverse proxy in front of the BFF shouldn't change that significantly. You probably just need to add the forwarded headers middleaAre to your BFF.

**Comment by greg-signi** [comment]
Yes there's no problem with that, my current solution is working. The only quirk my solution has is that I'm not hardcoding every API that needs to be redirected like you have on your samples with Yarp. 

(maybe I should have added this piece of the method controller that receives the destination api url in the full path)

```
[HttpGet("{**fullPath}")]
public async Task<IActionResult> Get()
{
    string? destinationApiUrl = HttpContext.GetRouteValue("fullPath")?.ToString();
    ...
    ...
```

So I **couldn't** use your extension methods `Duende.Bff.Yarp.RemoteApiEndpoint.Map(string localPath, string apiAddress).RequireAccessToken()` 

So I took some of the code from that method and came up with the solution I posted above.

And was just wondering if there could not be some other methods that could help my situation and avoid doing the proxy myself ?

**Comment by josephdecock** [comment]
You don't have to hard code every route in the BFF. If you're using YARP directly, you can add a /{**catch-all} to the end of the route, and if you're using our simplified router, it adds that automatically to the end of routes. Generally we think it is a good idea to have a good amount of control over which routes the BFF will make API calls to, so we create prefixes that go to specific APIs, but you can go as granular or specific as you like. And if you use full YARP, you can transform the routes as well (add or remove prefixes, do pattern matching etc). See [here]({https://microsoft.github.io/reverse-proxy/articles/transforms.html) for more info on transforms.
